codegen/proto/src/lib.rs

```diff
@@ -40,6 +40,179 @@ pub mod moby {
+                        send_compression_encodings: Default::default(),
+                        max_decoding_message_size: None,
+                        max_encoding_message_size: None,
+                    }
+                }
+                pub fn with_interceptor<F>(inner: T, interceptor: F) -> InterceptedService<Self, F>
+                where
+                    F: tonic::service::Interceptor,
+                {
+                    InterceptedService::new(Self::new(inner), interceptor)
+                }
+                /// Enable decompressing requests with the given encoding.
+                #[must_use]
+                pub fn accept_compressed(mut self, encoding: CompressionEncoding) -> Self {
+                    self.accept_compression_encodings.enable(encoding);
+                    self
+                }
+                /// Compress responses with the given encoding, if the client supports it.
+                #[must_use]
+                pub fn send_compressed(mut self, encoding: CompressionEncoding) -> Self {
+                    self.send_compression_encodings.enable(encoding);
+                    self
+                }
+                /// Limits the maximum size of a decoded message.
+                ///
+                /// Default: `4MB`
+                #[must_use]
+                pub fn max_decoding_message_size(mut self, limit: usize) -> Self {
+                    self.max_decoding_message_size = Some(limit);
+                    self
+                }
+                /// Limits the maximum size of an encoded message.
+                ///
+                /// Default: `usize::MAX`
+                #[must_use]
+                pub fn max_encoding_message_size(mut self, limit: usize) -> Self {
+                    self.max_encoding_message_size = Some(limit);
+                    self
+                }
+            }
+            impl<T, B> tonic::codegen::Service<http::Request<B>> for FileSendPacketServer<T>
+            where
+                T: FileSendPacket,
+                B: Body + Send + 'static,
+                B::Error: Into<StdError> + Send + 'static,
+            {
+                type Response = http::Response<tonic::body::BoxBody>;
+                type Error = std::convert::Infallible;
+                type Future = BoxFuture<Self::Response, Self::Error>;
+                fn poll_ready(
+                    &mut self,
+                    _cx: &mut Context<'_>,
+                ) -> Poll<std::result::Result<(), Self::Error>> {
+                    Poll::Ready(Ok(()))
+                }
+                fn call(&mut self, req: http::Request<B>) -> Self::Future {
+                    let _inner = self.inner.clone();
+                    match req.uri().path() {
+                        "/moby.filesync.v1.FileSend/DiffCopy" => {
+                            #[allow(non_camel_case_types)]
+                            struct DiffCopySvc<T: FileSendPacket>(pub Arc<T>);
```

Comment:
I see the inner type is different here than the generated `FileSend` type. I guess that's what you need to send back to buildkit ?

Reply:
Yes, `crate::fsutil::types::Packet` is the correct message type for the gRPC stream! We both send and receive these packets when using the `local` exporter (as opposed to `BytesMessage` for the `tar` exporter).